### PR TITLE
Video Capabilities, Control Frame Payloads, and RTP encoder integration for SMM

### LINKED ISSUE #686: v5.0 - Automatically handle GetSystemCapabilities and Control Payload settings
### Enhancement
* v4.7 added the bare minimum to support [GetSystemCapabilities](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0055-system_capabilities_query.md) and [control payloads](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0078-control_frame_payloads_v1_0_0.md) in the StreamingMediaManager. We should automatically handle this for developers.

### PR DESCRIPTION
**Pull request by joeljfischer**
Fixes #686

This PR is **ready** for review.

### Risk
This PR makes **it's complicated** API changes.

The changes made are major, but strictly to already in-progress major changes.

This makes large changes to the setup flow of video streaming.

### Testing Plan
Tests will be added.

### Summary
This PR integrates a number of recent changes into the Streaming Media Manager.

Added support for video capabilities, video control frame payloads, and alters the lifecycle of the SMM.

##### Enhancements
* SDL-0058 (#628) added support for video capabilities. This manager will now ask the head unit for video capabilities using `GetSystemCapabilities`, and do the right thing if it returns or not. If it does, it will try to use the head unit's preferred settings and provides the app dev a way to modify them. If it does not, it will use H264 RAW with the display capabilities screen size.
* SDL-0052 / SDL-0078 (#616 / #663) The SMM will now use the negotiated preferred settings with a constructed video start service payload. If the head unit rejects this and NAKs, it will check the rejected parameters and retry with the next resolution or format (respective to the rejected parameters) and try again until the list is exhausted.
* Integrates the RTP encoder and will start it if H264 RTP is negotiated.

##### Bug Fixes
* The Streaming Media Manager is now not started by the lifecycle manager until `HMI_FULL` is reached.

### Tasks Remaining:
- [x] Unit Tests
- [x] Start RAW vs. RTP H264 encoders

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

File: SmartDeviceLink/SDLH264VideoEncoder.h [hunk 1]
```diff
@@ -10,22 +10,24 @@
...
-    SDLVideoEncoderErrorConfigurationCompressionSessionSetPropertyFailure = 2
+    SDLVideoEncoderErrorConfigurationCompressionSessionSetPropertyFailure = 2,
+    SDLVideoEncoderErrorProtocolUnknown = 3
 };
 
 extern NSString *const SDLErrorDomainVideoEncoder;
 
 
-@interface SDLVideoEncoder : NSObject
+@interface SDLH264VideoEncoder : NSObject
```
Comment (NicoleYarroch) [inline SmartDeviceLink/SDLH264VideoEncoder.h]: The class name also needs to be updated on line no. 2

File: SmartDeviceLink/SDLH264VideoEncoder.m [hunk 2]
```diff
@@ -18,21 +19,20 @@
 static NSDictionary<NSString *, id>* _defaultVideoEncoderSettings;
 
 
-@interface SDLVideoEncoder ()
+@interface SDLH264VideoEncoder ()
```
Comment (NicoleYarroch) [inline SmartDeviceLink/SDLH264VideoEncoder.m]: The class name also needs to be updated on line no. 2

File: SmartDeviceLink/SDLLifecycleManager.m [hunk 3]
```diff
@@ -274,19 +278,17 @@ - (void)didEnterStateSettingUpManagers {
...
-    }
-    
-    [self.streamManager startWithProtocol:self.proxy.protocol completionHandler:^(BOOL success, NSError * _Nullable error) {
-        if (!success) {
-            SDLLogE(@"Streaming media manager was unable to start; error: %@", error);
-        }
-        
-        dispatch_group_leave(managerGroup);
-    }];
+//    if (self.streamManager != nil) {
```
Comment (NicoleYarroch) [inline SmartDeviceLink/SDLLifecycleManager.m]: Delete this commented out code

File: SmartDeviceLink/SDLLifecycleManager.m [hunk 4]
```diff
@@ -452,6 +455,27 @@ + (BOOL)sdl_checkNotification:(NSNotification *)notification containsKindOfClass
     return YES;
 }
 
+- (void)sdl_onFirstHMIFull {
+    // Nothing yet
+}
+
+- (void)sdl_onFirstHMINonNone {
+    // If we are a nav / projection app and desire to stream, we need to be in HMI background, limited, or full and perform additional setup when that occurs
+    if (self.streamManager != nil) {
```
Comment (NicoleYarroch) [inline SmartDeviceLink/SDLLifecycleManager.m]: Just return right away if `(self.streamManager == nil)`

File: SmartDeviceLink/SDLLifecycleManager.m [hunk 5]
```diff
@@ -452,6 +455,27 @@ + (BOOL)sdl_checkNotification:(NSNotification *)notification containsKindOfClass
     return YES;
 }
 
+- (void)sdl_onFirstHMIFull {
+    // Nothing yet
```
Comment (NicoleYarroch) [inline SmartDeviceLink/SDLLifecycleManager.m]: Why is this function needed if nothing happens when it is called?

File: SmartDeviceLink/SDLLifecycleManager.m [hunk 6]
```diff
@@ -452,6 +455,27 @@ + (BOOL)sdl_checkNotification:(NSNotification *)notification containsKindOfClass
...
+        [self.streamManager startWithProtocol:self.proxy.protocol completionHandler:^(BOOL success, NSError * _Nullable error) {
+            if (!success) {
+                SDLLogE(@"Streaming media manager was unable to start; error: %@", error);
+                [weakSelf stop];
+            }
+
+            [weakSelf.lifecycleStateMachine transitionToState:SDLLifecycleStateReady];
+        }];
+
+        return;
```
Comment (NicoleYarroch) [inline SmartDeviceLink/SDLLifecycleManager.m]: What is the purpose of this `return`?

File: SmartDeviceLink/SDLLifecycleManager.m [hunk 7]
```diff
@@ -104,6 +106,8 @@ - (instancetype)initWithConfiguration:(SDLConfiguration *)configuration delegate
     _notificationDispatcher = [[SDLNotificationDispatcher alloc] init];
     _responseDispatcher = [[SDLResponseDispatcher alloc] initWithNotificationDispatcher:_notificationDispatcher];
     _registerResponse = nil;
+    _firstHMIFullOccurred = NO;
+    _firstHMINonNoneOccurred = NO;
```
Comment (NicoleYarroch) [inline SmartDeviceLink/SDLLifecycleManager.m]: `firstHMIFullOccurred` and `firstHMINonNoneOccurred` never get set to `YES`

File: SmartDeviceLink/SDLRAWH264Packetizer.h [hunk 8]
```diff
@@ -7,11 +7,12 @@
 //
 
 #import <Foundation/Foundation.h>
+
 #import "SDLH264Packetizer.h"
 
 NS_ASSUME_NONNULL_BEGIN
 
-@interface SDLH264ByteStreamPacketizer : NSObject <SDLH264Packetizer>
+@interface SDLRAWH264Packetizer : NSObject <SDLH264Packetizer>
```
Comment (NicoleYarroch) [inline SmartDeviceLink/SDLRAWH264Packetizer.h]: Class name also needs to get updated on line no. 2

File: SmartDeviceLink/SDLRAWH264Packetizer.m [hunk 9]
```diff
@@ -7,15 +7,16 @@
...
 
 #import <Foundation/Foundation.h>
-#import "SDLH264ByteStreamPacketizer.h"
+#import "SDLRAWH264Packetizer.h"
+#import "SDLH264Packetizer.h"
 
 NS_ASSUME_NONNULL_BEGIN
 
-@interface SDLH264ByteStreamPacketizer ()
+@interface SDLRAWH264Packetizer ()
```
Comment (NicoleYarroch) [inline SmartDeviceLink/SDLRAWH264Packetizer.m]: The class name also needs to get updated on line no. 2

File: SmartDeviceLink/SDLStreamingMediaLifecycleManager.m [hunk 10]
```diff
@@ -103,38 +126,65 @@ - (instancetype)initWithEncryption:(SDLStreamingEncryptionFlag)encryption videoE
...
     _lastPresentationTimestamp = kCMTimeInvalid;
 
     return self;
 }
 
-- (void)startWithProtocol:(SDLAbstractProtocol *)protocol completionHandler:(void (^)(BOOL, NSError * _Nullable))completionHandler {
+- (void)startWithProtocol:(SDLAbstractProtocol *)protocol completionHandler:(SDLStreamingMediaReadyBlock)readyHandler {
     _protocol = protocol;
-    
+    _readyHandler = readyHandler;
```
Comment (NicoleYarroch) [inline SmartDeviceLink/SDLStreamingMediaLifecycleManager.m]: It doesn't look like the `readyHandler` is ever used

File: SmartDeviceLink/SDLStreamingMediaLifecycleManager.m [hunk 11]
```diff
@@ -371,40 +416,121 @@ - (void)didEnterStateAudioStreamShuttingDown {
...
+    [self sdl_transitionToStoppedState:endServiceNAK.header.serviceType];
 }
 
 #pragma mark - SDLVideoEncoderDelegate
-- (void)videoEncoder:(SDLVideoEncoder *)encoder hasEncodedFrame:(NSData *)encodedVideo {
-    SDLLogV(@"Video encoder encoded frame, sending");
+
+- (void)videoEncoder:(SDLH264VideoEncoder *)encoder hasEncodedFrame:(NSData *)encodedVideo {
+    SDLLogV(@"Video encoder encoded frame, sending data");
     // Do we care about app state here? I don't think so…
```
Comment (NicoleYarroch) [inline SmartDeviceLink/SDLStreamingMediaLifecycleManager.m]: Comment can be deleted

File: SmartDeviceLink/SDLStreamingMediaLifecycleManager.m (hunk 10 above)
Comment (NicoleYarroch) [inline SmartDeviceLink/SDLStreamingMediaLifecycleManager.m]: The ready handler is never called.

File: SmartDeviceLinkTests/SDLStreamingMediaLifecycleManagerSpec.m [hunk 12]
```diff
@@ -7,59 +7,87 @@
...
+        expect(streamingLifecycleManager.videoFormat).to(beNil());
+        expect(streamingLifecycleManager.dataSource).to(equal(testDataSource));
+        expect(streamingLifecycleManager.supportedFormats).to(haveCount(2));
+        expect(streamingLifecycleManager.preferredFormats).to(beNil());
+        expect(streamingLifecycleManager.preferredResolutions).to(beNil());
+        expect(streamingLifecycleManager.preferredFormatIndex).to(equal(0));
+        expect(streamingLifecycleManager.preferredResolutionIndex).to(equal(0));
     });
     
     describe(@"when started", ^{
```
Comment (NicoleYarroch) [inline SmartDeviceLinkTests/SDLStreamingMediaLifecycleManagerSpec.m]: If you put a breakpoint inside the `startWithProtocol:protocol:completionHandler` method you can see that completion handler is never called

```
 [streamingLifecycleManager startWithProtocol:protocolMock completionHandler:^(BOOL success, NSError * _Nullable error) {
                readyHandlerSuccess = success;
                readyHandlerError = error;
            }];
```

**Comment by joeljfischer** [comment]
@NicoleYarroch Review comment addressed

**Comment by NicoleYarroch** [comment]
The test cases are not compiling due to changes to the `startWithProtocol` method

**Comment by joeljfischer** [comment]
@NicoleYarroch Fixed

**Pull request merged**